cache/cache.go

```diff
@@ -0,0 +1,185 @@
+// The cache package provides a concurrency-safe in-memory caching system designed for storing and managing
+// hierarchical cache data related to cloud environments, services, and metric families. The system
+// is designed around a singleton pattern, ensuring only one instance of the cache backend is created and used
+// throughout the application. It supports operations for setting and retrieving cache data at different levels
+// (cloud, service, and metric family), with each level structured to contain relevant data and timestamps for
+// cache validity checks. Additionally, the cache provides functionality to flush data based on age (TTL) and
+// to clear the entire cache.
+
+package cache
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	dto "github.com/prometheus/client_model/go"
+)
+
+var singleCache CacheBackend
+var once sync.Once
+
+type CacheBackend interface {
+	// Set cache with cloud name.
+	SetCloudCache(string, CloudCache)
+	// Get cache with cloud name.
+	GetCloudCache(string) (CloudCache, bool)
+	// Set cache with service name.
+	SetServiceCache(string, string, ServiceCache)
```

Comment:
I would like to keep those methods. Let's see other's feedback.

Reply:
@jneo8 If unused, please remove.

Reply:
They are used in the unit-tests.
Remove some of them will make the code very hard to test.

For example if I want to test `SetMetricFamilyCache`, I need to have a work-around way to get the cache object from maybe server or cloud level then verify it's exists, which also make testing code really dirty.

Also if you remove `GetCloudCache`, it will somehow remove the work-around way also.

Since those functions are just get/set methods and either been used in the code or test code. I would like to keep them in the interface.

Reply:
The refactor is done in the [newest commit](https://github.com/openstack-exporter/openstack-exporter/pull/355/commits/b89be8c183a77b71a39c5f08654daf59beda23b2)